toolchain/check/handle_function.cpp

```diff
@@ -182,15 +242,100 @@ static auto TryMergeRedecl(Context& context, Parse::AnyFunctionDeclId node_id,
   }
 
   if (!prev_function_id.has_value()) {
-    DiagnoseDuplicateName(context, name_id, name_loc_id, prev_id);
+    DiagnoseDuplicateName(context, name_context.name_id, name_context.loc_id,
+                          prev_id);
     return;
   }
 
   if (MergeFunctionRedecl(context, node_id, function_info, is_definition,
                           prev_function_id, prev_import_ir_id)) {
     // When merging, use the existing function rather than adding a new one.
     function_decl.function_id = prev_function_id;
+    function_decl.type_id = prev_type_id;
+  }
+}
+
+// Adds the declaration to name lookup when appropriate.
+static auto MaybeAddToNameLookup(
+    Context& context, const DeclNameStack::NameContext& name_context,
+    const KeywordModifierSet& modifier_set,
+    const std::optional<SemIR::Inst>& parent_scope_inst, SemIR::InstId decl_id)
+    -> void {
+  if (name_context.state == DeclNameStack::NameContext::State::Poisoned ||
+      name_context.prev_inst_id().has_value()) {
+    return;
+  }
+
+  // At interface scope, a function declaration introduces an associated
+  // function.
+  auto lookup_result_id = decl_id;
+  if (parent_scope_inst && !name_context.has_qualifiers) {
+    if (auto interface_scope =
+            parent_scope_inst->TryAs<SemIR::InterfaceDecl>()) {
+      lookup_result_id = BuildAssociatedEntity(
+          context, interface_scope->interface_id, decl_id);
+    }
   }
+
+  context.decl_name_stack().AddName(name_context, lookup_result_id,
+                                    modifier_set.GetAccessKind());
+}
+
+// If the function is the entry point, do corresponding validation.
+static auto ValidateForEntryPoint(Context& context,
+                                  Parse::AnyFunctionDeclId node_id,
+                                  SemIR::FunctionId function_id,
+                                  const SemIR::Function& function_info)
+    -> void {
+  if (!SemIR::IsEntryPoint(context.sem_ir(), function_id)) {
+    return;
+  }
+
+  auto return_type_id = function_info.GetDeclaredReturnType(context.sem_ir());
+  // TODO: Update this once valid signatures for the entry point are decided.
+  if (function_info.implicit_param_patterns_id.has_value() ||
+      !function_info.param_patterns_id.has_value() ||
+      !context.inst_blocks().Get(function_info.param_patterns_id).empty() ||
+      (return_type_id.has_value() &&
+       return_type_id != GetTupleType(context, {}) &&
+       // TODO: Decide on valid return types for `Main.Run`. Perhaps we should
+       // have an interface for this.
+       return_type_id != MakeIntType(context, node_id, SemIR::IntKind::Signed,
+                                     context.ints().Add(32)))) {
```

Comment:
Might be easier to read as a series of early returns?
```
if (!function_info.implicit_param_patterns_id.has_value()) {
  return;
}
if (function_info.param_patterns_id.has_value()) {
  return;
}
if (context.inst_blocks().Get(function_info.param_patterns_id).empty()) {
  return;
}
if (!return_type_id.has_value()) {
  return;
}
if (return_type_id == GetTupleType(context, {})) {
  return;
}
if (return_type_id == MakeIntType(..)) {
  return;
}
CARBON_DIAGNOSTIC...
```

Reply:
I lean towards leaving this as-is. I think the structure is okay to read... but also I think it should probably be doing different validation. There are already TODOs to revisit this though, which is part of why I hold back here.